src/python/pants/core/goals/typecheck.py

```diff
@@ -0,0 +1,130 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import itertools
+from dataclasses import dataclass
+from typing import Iterable
+
+from pants.core.goals.style_request import StyleRequest
+from pants.core.util_rules.filter_empty_sources import (
+    FieldSetsWithSources,
+    FieldSetsWithSourcesRequest,
+)
+from pants.engine.collection import Collection
+from pants.engine.console import Console
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.process import FallibleProcessResult
+from pants.engine.rules import goal_rule
+from pants.engine.selectors import Get, MultiGet
+from pants.engine.target import TargetsWithOrigins
+from pants.engine.unions import UnionMembership, union
+from pants.util.strutil import strip_v2_chroot_path
+
+
+@dataclass(frozen=True)
+class TypecheckResult:
+    exit_code: int
+    stdout: str
+    stderr: str
+    typechecker_name: str
+
+    @staticmethod
+    def from_fallible_process_result(
+        process_result: FallibleProcessResult,
+        *,
+        typechecker_name: str,
+        strip_chroot_path: bool = False,
+    ) -> "TypecheckResult":
+        def prep_output(s: bytes) -> str:
+            return strip_v2_chroot_path(s) if strip_chroot_path else s.decode()
+
+        return TypecheckResult(
+            exit_code=process_result.exit_code,
+            stdout=prep_output(process_result.stdout),
+            stderr=prep_output(process_result.stderr),
+            typechecker_name=typechecker_name,
+        )
+
+
+class TypecheckResults(Collection[TypecheckResult]):
+    """Zero or more TypecheckResult objects for a single type checker.
+
+    Typically, type checkers will return one result. If they no-oped, they will return zero results.
+    However, some type checkers may need to partition their input and thus may need to return
+    multiple results.
+    """
```

Comment:
There's no need yet to batch things. We expect MyPy to return one single result. But, it's very very little extra code for us to have type checkers return a collection, and I think we'll appreciate the flexibility in the future.